scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -545,27 +555,28 @@ object Execution {
   }
 
   private case class Zipped[S, T](one: Execution[S], two: Execution[T]) extends Execution[(S, T)] {
-    def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) =
-      cache.getOrElseInsert(conf, this, {
-        val f1 = one.runStats(conf, mode, cache)
-        val f2 = two.runStats(conf, mode, cache)
-        failFastZip(f1, f2)
-          .map { case ((s, ss), (t, st)) => ((s, t), Monoid.plus(ss, st)) }
-      })
+    protected def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) =
+      for {
+        f1 <- Trampoline.call(one.runStats(conf, mode, cache))
+        f2 <- Trampoline.call(two.runStats(conf, mode, cache))
+      } yield {
+        cache.getOrElseInsert(conf, this,
+          failFastZip(f1, f2)
+            .map { case ((s, ss), (t, st)) => ((s, t), Monoid.plus(ss, st)) })
+      }
   }
   private case class UniqueIdExecution[T](fn: UniqueID => Execution[T]) extends Execution[T] {
-    def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) =
-      cache.getOrElseInsert(conf, this, {
-        val (uid, nextConf) = conf.ensureUniqueId
-        fn(uid).runStats(nextConf, mode, cache)
-      })
+    protected def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) = {
+      val (uid, nextConf) = conf.ensureUniqueId
+      Trampoline.call(fn(uid).runStats(nextConf, mode, cache))
```

Comment:
Ah, of course.  I guess it would be possible then to still blow the stack via nested `withId`.  That could be addressed by having an ID map/cache, but maybe that's overkill for a relatively contrived use case?